src/modules/control_allocator/ControlAllocation/ControlAllocationPseudoInverse.cpp

```diff
@@ -44,44 +44,65 @@
 void
 ControlAllocationPseudoInverse::setEffectivenessMatrix(
 	const matrix::Matrix<float, ControlAllocation::NUM_AXES, ControlAllocation::NUM_ACTUATORS> &effectiveness,
-	const ActuatorVector &actuator_trim, const ActuatorVector &linearization_point, int num_actuators)
+	const ActuatorVector &actuator_trim, const ActuatorVector &linearization_point, int num_actuators,
+	bool update_normalization_scale)
 {
-	ControlAllocation::setEffectivenessMatrix(effectiveness, actuator_trim, linearization_point, num_actuators);
+	ControlAllocation::setEffectivenessMatrix(effectiveness, actuator_trim, linearization_point, num_actuators,
+			update_normalization_scale);
 	_mix_update_needed = true;
+	_normalization_needs_update = update_normalization_scale;
 }
 
 void
 ControlAllocationPseudoInverse::updatePseudoInverse()
 {
 	if (_mix_update_needed) {
 		matrix::geninv(_effectiveness, _mix);
+
+		if (_normalization_needs_update) {
+			updateControlAllocationMatrixScale();
+			_normalization_needs_update = false;
+		}
+
 		normalizeControlAllocationMatrix();
 		_mix_update_needed = false;
 	}
 }
 
 void
-ControlAllocationPseudoInverse::normalizeControlAllocationMatrix()
+ControlAllocationPseudoInverse::updateControlAllocationMatrixScale()
 {
+	// Same scale on roll and pitch
 	if (_normalize_rpy) {
-		// Same scale on roll and pitch
-		const float roll_norm_sq = _mix.col(0).norm_squared();
-		const float pitch_norm_sq = _mix.col(1).norm_squared();
-		_control_allocation_scale(0) = sqrtf(fmaxf(roll_norm_sq, pitch_norm_sq) / (_num_actuators / 2.f));
-		_control_allocation_scale(1) = _control_allocation_scale(0);
 
-		if (_control_allocation_scale(0) > FLT_EPSILON) {
-			_mix.col(0) /= _control_allocation_scale(0);
-			_mix.col(1) /= _control_allocation_scale(1);
+		int num_non_zero_roll_torque = 0;
+		int num_non_zero_pitch_torque = 0;
+
+		for (int i = 0; i < _num_actuators; i++) {
+
+			if (fabsf(_mix(i, 0)) > 1e-3f) {
+				++num_non_zero_roll_torque;
+			}
+
+			if (fabsf(_mix(i, 1)) > 1e-3f) {
+				++num_non_zero_pitch_torque;
+			}
 		}
 
-		// Scale yaw separately
-		_control_allocation_scale(2) = _mix.col(2).max();
+		float roll_norm_scale = 1.f;
+		float pitch_norm_scale = 1.f;
 
-		if (_control_allocation_scale(2) > FLT_EPSILON) {
-			_mix.col(2) /= _control_allocation_scale(2);
+		if (num_non_zero_roll_torque > 0) {
+			roll_norm_scale = sqrtf(_mix.col(0).norm_squared() / (num_non_zero_roll_torque / 2.f));
+			pitch_norm_scale = sqrtf(_mix.col(1).norm_squared() / (num_non_zero_pitch_torque / 2.f));
```

Comment:
This should be checked separately: `if (num_non_zero_pitch_torque > 0) {`

Reply:
yes ofc, thanks for raising, pushed a fix.